docs/pipeline.md

```diff
@@ -73,30 +73,36 @@ where what the phases of Sorbet actually do are documented.
- ┌────────────────────────────────────────────────────────────────────────────────┐
- │  ┌───────────────────┐                                                         │
- │  │    GlobalState    │            typecheck                                    │
- │  └───────────────────┘                                                         │
- ├──────────────────────────┬──────────────────────────┬──────────────────────────┤
- │                          │                          │                          │
- │  ┌───────────────────┐   │   ┌───────────────────┐  │   ┌───────────────────┐  │
- │  │  ast::Expression  │   │   │  ast::Expression  │  │   │  ast::Expression  │  │
- │  └───────────────────┘   │   └───────────────────┘  │   └───────────────────┘  │
- │            │             │             │            │             │            │
- │            │ cfg         │             │ cfg        │             │ cfg        │
- │            ▼             │             ▼            │             ▼            │
- │      ┌──────────┐        │       ┌──────────┐       │       ┌──────────┐       │
- │      │ cfg::CFG │        │       │ cfg::CFG │       │       │ cfg::CFG │       │
- │      └──────────┘        │       └──────────┘       │       └──────────┘       │
- │            │             │             │            │             │            │
- │            │ infer       │             │ infer      │             │ infer      │
- │            ▼             │             ▼            │             ▼            │
- │   ┌─────────────────┐    │    ┌─────────────────┐   │    ┌─────────────────┐   │
- │   │    cfg::CFG     │    │    │    cfg::CFG     │   │    │    cfg::CFG     │   │
- │   │  (with types)   │    │    │  (with types)   │   │    │  (with types)   │   │
- │   └─────────────────┘    │    └─────────────────┘   │    └─────────────────┘   │
- │                          │                          │                          │
- └──────────────────────────┴──────────────────────────┴──────────────────────────┘
+ ┌──────────────────────────────────────────────────────────────────────────────────┐
+ │  ┌───────────────────┐                                                           │
+ │  │    GlobalState    │            typecheck                                      │
+ │  └───────────────────┘                                                           │
+ ├──────────────────────────┬───────────────────────────┬───────────────────────────┤
+ │                          │                           │                           │
+ │  ┌───────────────────┐   │   ┌───────────────────┐   │   ┌───────────────────┐   │
+ │  │  ast::Expression  │   │   │  ast::Expression  │   │   │  ast::Expression  │   │
+ │  └───────────────────┘   │   └───────────────────┘   │   └───────────────────┘   │
+ │   │                      │    │                      │    │                      │
+ │   │ definition_validator │    │ definition_validator │    │ definition_validator │
```

Comment:
is flatten run right after this one? You could put it here? https://github.com/sorbet/sorbet/blob/74acf3d4/main/pipeline/pipeline.cc#L677

Reply:
Flatten is run in two places right now from looking at the code. @aisamanra why do we have flatten in two places?

Reply:
They do slightly different things, and every time I tried to remove the later psas there was a problem with the PR and it got reverted. In effect, the first one makes sure that no method definitions are nested, while the second one makes sure that all method definitions are top-level.

Reply:
To be clear: the goal is indeed to remove the later one, but there's a lot more subtlety there than you might expect. I don't think I have a document explaining why, but I should write one. It's worth noting that the second flattening pass also won't go away entirely, because it's also the `static-init`-generation pass, so there will still be a pass there, it's just not responsible for flattening any more.